### DISCUSSION TITLE
Shovel not respecting prefetch for "no-ack" and "on-publish"?

### DISCUSSION THREAD

**Discussion by johanrhodin** [rabbitmq/rabbitmq-server#3407] [Questions] [answered]
Consider the following scenario, two clusters connected by a shovel. The shovel is using these parameters:
```
    "ack-mode": "on-publish",
    "src-prefetch-count": 1
```

<img width="486" alt="Overview of configuration with two clusters connected via a shovel" src="https://user-images.githubusercontent.com/460369/132930542-202ee5d1-0bb0-4068-89f0-ec8e65dad51e.png">

If the destination cluster reaches the high watermark the messages at the source cluster will accumulate. Unfortunately they don't accumulate in the source queue but in the connection writer. 

This can be seen in the Erlang mailbox for the `rabbit_writer` that keeps on growing.
<img width="973" alt="rabbit_writer enter_mainloop" src="https://user-images.githubusercontent.com/460369/132930648-05c8bc22-1027-41aa-b65d-73b3fd53fa9a.png">

The exact shovel(s) this happens to can be found by using the technique outlined here: https://groups.google.com/g/rabbitmq-users/c/bJqiqyFuiUE/m/tJNZaUKyBwAJ

The only workaround that I've found for this is to use ack-mode on-confirm, but it would be good to have a way for less important data (that do not require a confirm) to not accumulate in the RAM of the sending node.

If this is considered to be "by design" I'd be happy to include that in the documentation for shovel. This issue comes up from time to time and is hard to reason about and to track down.

A more detailed reproduction can be seen in this gist: https://gist.github.com/johanrhodin/87a54361ddebc2c09c015d991385e329

**Comment by carlhoerberg** [comment]
How come erlang-rabbitmq-client can have an unbounded write buffer? `connection_writer`'s mailbox seems to just grow and grow when the server is blocking publishes (not reading from its side of the socket).

**Comment by johanrhodin** [comment, marked as answer]
This is being addressed in https://github.com/rabbitmq/rabbitmq-server/pull/5715
